File: python/mirage/commands/builtin/generic/sed.py
import re
from collections.abc import AsyncIterator, Awaitable, Callable

from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.sed_helper import (_execute_program,
                                                _parse_one_command,
                                                _parse_program)
from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _is_simple_sub(commands: list[dict], suppress: bool) -> bool:
    return (len(commands) == 1 and commands[0]["cmd"] == "s"
            and commands[0].get("addr_start") is None and not suppress)


async def sed(
    paths: list[PathSpec],
    expression: str,
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    write_bytes: Callable[..., Awaitable[None]],
    accessor: object = None,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    in_place: bool = False,
    suppress: bool = False,
    index: IndexCacheStore | None = None,
) -> tuple[ByteSource | None, IOResult]:
    if ";" in expression or "{" in expression:
        commands = _parse_program(expression)
    else:
        commands = [_parse_one_command(expression)[0]]

    if paths and _is_simple_sub(commands, suppress):
        parsed = commands[0]
        re_flags = re.IGNORECASE if "i" in parsed["expr_flags"] else 0
        count = 0 if "g" in parsed["expr_flags"] else 1

        if in_place:
            writes: dict[str, bytes] = {}
            for p in paths:
                data = await read_bytes(accessor, p)
                text = data.decode(errors="replace")
                new_text = re.sub(parsed["pattern"],
                                  parsed["replacement"],
                                  text,
                                  count=count,
                                  flags=re_flags)
                new_data = new_text.encode()
                await write_bytes(accessor, p, new_data)
                writes[p.original] = new_data
            return None, IOResult(writes=writes,
                                  cache=[p.original for p in paths])

        outputs: list[str] = []
        for p in paths:
            data = await read_bytes(accessor, p)
            text = data.decode(errors="replace")
            new_text = re.sub(parsed["pattern"],
                              parsed["replacement"],
                              text,
                              count=count,
                              flags=re_flags)
            outputs.append(new_text)
        return "".join(outputs).encode(), IOResult(
            cache=[p.original for p in paths])

    if paths:
        modifying = in_place and any(c["cmd"] in ("s", "d") for c in commands)
        all_outputs: list[str] = []
        writes = {}
        for p in paths:
            data = await read_bytes(accessor, p)
            text = data.decode(errors="replace")
            result = _execute_program(text, commands, suppress=suppress)
            if modifying:
                new_data = result.encode()
                await write_bytes(accessor, p, new_data)
                writes[p.original] = new_data
            else:
                all_outputs.append(result)
        if modifying:
            return None, IOResult(writes=writes,
                                  cache=[p.original for p in paths])
        return "\n".join(all_outputs).encode(), IOResult()

    raw = await _read_stdin_async(stdin)
    if raw is None:
        raise ValueError("sed: usage: sed EXPRESSION path")
    text = raw.decode(errors="replace")
    result = _execute_program(text, commands, suppress=suppress)
    return result.encode(), IOResult()


__all__ = ["sed"]

File: typescript/packages/core/src/commands/builtin/generic/sed.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import {
  executeProgram,
  parseOneCommand,
  parseProgram,
  translateReplacement,
  type SedCommand,
} from '../sed_helper.ts'
import { readStdinAsync } from '../utils/stream.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

export async function sedGeneric(
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
  write: (p: PathSpec, data: Uint8Array) => Promise<void>,
): Promise<CommandFnResult> {
  const script = texts[0]
  if (script === undefined) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('sed: missing script\n') })]
  }
  const suppress = opts.flags.n === true
  const inPlace = opts.flags.i === true
  let commands: SedCommand[]
  try {
    if (script.includes(';') || script.includes('{')) {
      commands = parseProgram(script)
    } else {
      commands = [parseOneCommand(script)[0]]
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
  }
  const first = commands[0]
  const isSimpleSub =
    commands.length === 1 &&
    first?.cmd === 's' &&
    (first.addrStart === null || first.addrStart === undefined) &&
    !suppress

  if (paths.length > 0) {
    if (isSimpleSub) {
      const pat = first.pattern ?? ''
      const repl = translateReplacement(first.replacement ?? '')
      const ef = first.exprFlags ?? ''
      const ignoreCase = ef.includes('i')
      const global = ef.includes('g')
      const flags = (ignoreCase ? 'i' : '') + (global ? 'g' : '')
      if (inPlace) {
        const writes: Record<string, Uint8Array> = {}
        for (const p of paths) {
          const data = await materialize(stream(p))
          const text = DEC.decode(data)
          const newText = text.replace(new RegExp(pat, flags), repl)
          const newData = ENC.encode(newText)
          await write(p, newData)
          writes[p.stripPrefix] = newData
        }
        return [null, new IOResult({ writes, cache: paths.map((p) => p.stripPrefix) })]
      }
      const outputs: string[] = []
      for (const p of paths) {
        const data = await materialize(stream(p))
        const text = DEC.decode(data)
        outputs.push(text.replace(new RegExp(pat, flags), repl))
      }
      const out: ByteSource = ENC.encode(outputs.join(''))
      return [out, new IOResult({ cache: paths.map((p) => p.stripPrefix) })]
    }

    const modifying = inPlace && commands.some((c) => c.cmd === 's' || c.cmd === 'd')
    const allOutputs: string[] = []
    const writes: Record<string, Uint8Array> = {}
    for (const p of paths) {
      const data = await materialize(stream(p))
      const text = DEC.decode(data)
      const result = executeProgram(text, commands, suppress)
      if (modifying) {
        const newData = ENC.encode(result)
        await write(p, newData)
        writes[p.stripPrefix] = newData
      } else {
        allOutputs.push(result)
      }
    }
    if (modifying) {
      return [null, new IOResult({ writes, cache: paths.map((p) => p.stripPrefix) })]
    }
    const out: ByteSource = ENC.encode(allOutputs.join('\n'))
    return [out, new IOResult()]
  }

  const raw = await readStdinAsync(opts.stdin)
  if (raw === null) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('sed: missing operand\n') })]
  }
  const text = DEC.decode(raw)
  const result = executeProgram(text, commands, suppress)
  return [ENC.encode(result), new IOResult()]
}
